discovery/discovery.go

```diff
@@ -221,7 +221,7 @@ func (d *discovery) checkCluster() ([]*client.Node, int, uint64, error) {
 		if err == client.ErrInvalidJSON {
 			return nil, 0, 0, ErrBadDiscoveryEndpoint
 		}
-		if err == context.DeadlineExceeded {
+		if _, ok := err.(*client.ClusterError); ok {
```

Comment:
If it uses `http://wrong.discovery.io/` as discovery service, it will fail with `dial tcp: lookup wrong.discovery.io: no such host` after 15s. This branch makes it retry forever. It is a DNS failure, and keep retrying doesn't help.

I think if it is a timeout(either net or context), it is reasonable to do retry because we cannot tell what is happening. It needs verbosal error printout to help user debug, considering that it always give `connection to xxx timed out` today.